### PR TITLE
Add GraphQL as separate protocol

### PR DESCRIPTION
**Pull request by svkirillov**

### CODE REVIEW

File: internal/report/chart.go [hunk 1]
```diff
@@ -146,5 +152,15 @@ func generateChartData(s *db.Statistics) (
...
+	// Fix label for GraphQL if it is unavailable
+	if !s.IsGraphQLAvailable && containsApiCat {
+		for i := 0; i < len(apiIndicators); i++ {
+			if strings.HasPrefix(apiIndicators[i], "graphql") {
+				apiIndicators[i] = "graphql (unavailable)"
+				apiItems[i] = float32(0)
+			}
+		}
+	}
+
```
Comment (manmolecular) [inline internal/report/chart.go:164]: This block:
```
// Fix label for GraphQL if it is unavailable
...
```
And the block above:
```
// Fix label for gRPC if it is unavailable
...
```
Looks almost identical. Maybe it is a good idea to think about how this logic can be generalized into the function. As for now, it is ok (only 2 occurences), but in the future, if we want to use this primitive 3, 4, 5 ... times, it is better to think about function representation.

File: internal/scanner/graphql.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+	if ok {
+		endpointURL.RawQuery = ""
+
+		s.cfg.GraphQlURL = endpointURL.String()
+		s.httpClient.isGraphQlAvailable = true
+
+		return true, nil
+	}
+
+	return false, nil
```
Comment (manmolecular) [inline internal/scanner/graphql.go:129]: Copy-paste of the cycle above, symbol to symbol. Don't we want to make it a function or smthng?

File: internal/scanner/http_client.go [hunk 3]
```diff
@@ -81,18 +83,19 @@ func NewHTTPClient(cfg *config.Config) (*HTTPClient, error) {
...
+		renewSession:       cfg.RenewSession,
+		isGraphQlAvailable: true,
 	}, nil
 }
 
 func (c *HTTPClient) SendPayload(
 	ctx context.Context,
 	targetURL, placeholderName, encoderName, payload string,
-	testHeaderValue string,
+	degubHeaderValue string,
```
Comment (manmolecular) [inline internal/scanner/http_client.go:98]: What is `degub`? Is it on purpose or is it a typo? `debug`? `debugHeaderValue`?

File: internal/scanner/scanner.go [hunk 4]
```diff
@@ -119,6 +119,32 @@ func (s *Scanner) CheckGRPCAvailability(ctx context.Context) {
...
+			"status":     "done",
+			"connection": "not available",
+		}).WithError(err).Infof("GraphQL pre-check")
+	}
+	if available {
+		s.logger.WithFields(logrus.Fields{
+			"status":     "done",
+			"connection": "available",
+		}).Info("GraphQL pre-check")
+	} else {
```
Comment (manmolecular) [inline internal/scanner/scanner.go:138]: Can't we simplify this tree?
Something like:
1. if err != nil or not available = "not available"
2. else (... if err is nil and it is available) = "available"

Comment (manmolecular) [reply internal/scanner/scanner.go:138]: These two blocks (error or not available) are the same basically

File: cmd/flags.go [hunk 5]
```diff
@@ -152,14 +153,43 @@ func parseFlags() (args string, err error) {
...
+
+	// format GraphQL URL from given HTTP URL
+	if *graphqlURL == "" {
+		gqlValidURL := *validURL
+
+		gqlValidURL.Path = ""
+
+		*graphqlURL = gqlValidURL.String()
+	} else {
+		gqlValidURL, err := url.Parse(*wsURL)
```
[the remark this answers is not in the document]
Comment (svkirillov) [reply cmd/flags.go]: Thanks! Fixed - https://github.com/wallarm/gotestwaf/pull/167/commits/0363e8a30a5742e22da15fd2dc0aa5815a5a0b0b

File: internal/report/chart.go (hunk 1 above)
Comment (svkirillov) [reply internal/report/chart.go]: Fixed - https://github.com/wallarm/gotestwaf/pull/167/commits/60bf140e662611fb65a6840d0fba1092805756f5

File: internal/scanner/graphql.go (hunk 2 above)
Comment (svkirillov) [reply internal/scanner/graphql.go]: Fixed - https://github.com/wallarm/gotestwaf/pull/167/commits/447b4232abee113a0543330e9029ab4582b91837

File: internal/scanner/http_client.go (hunk 3 above)
Comment (svkirillov) [reply internal/scanner/http_client.go]: Fixed - https://github.com/wallarm/gotestwaf/pull/167/commits/8a90e824a59e77cb4781533c2d6b78cf34543a5a

File: internal/scanner/scanner.go (hunk 4 above)
Comment (svkirillov) [reply internal/scanner/scanner.go]: Fixed - https://github.com/wallarm/gotestwaf/pull/167/commits/30ea977db6ab0bd30606474a186f16a33b012a9f

File: cmd/flags.go [hunk 6]
```diff
@@ -152,14 +153,43 @@ func parseFlags() (args string, err error) {
...
+	}
+
+	// format GraphQL URL from given HTTP URL
+	if *graphqlURL == "" {
+		gqlValidURL := *validURL
+
+		gqlValidURL.Path = ""
+
+		*graphqlURL = gqlValidURL.String()
+	} else {
```
[the remark this answers is not in the document]
Comment (svkirillov) [reply cmd/flags.go]: Fixed - https://github.com/wallarm/gotestwaf/pull/167/commits/95925d673e7f91e375ee8591a4ae3eb7daa8c472

**Review by manmolecular — APPROVED** [review]

**Pull request merged**